src/commons/eventsourcing/kv.rs

```diff
@@ -341,14 +341,39 @@ impl KeyValueStoreDiskImpl {
         let path_str = path.to_string_lossy().into_owned();
 
         if path.exists() {
-            let f = File::open(path).map_err(|e| {
-                KrillIoError::new(
-                    format!("Could not read value for key '{}' from file '{}'", key, path_str),
-                    e,
-                )
-            })?;
-            let v = serde_json::from_reader(f)?;
-            Ok(Some(v))
+            // We read the json file into memory first. Deserializing from a slice is
+            // about 50-100 times faster than if we use serde_json::from_reader on the file.
+            //
+            // We could use a BufReader, but this is still a whole lot slower for large files.
+            // Based on our testing about 20 times slower for a 450 MB json file. The test
+            // and large file are not checked in to avoid that checkouts of this code base
+            // take up more space than needed, but it was tested using the pubd::RepositoryContent
+            // from a benchmark server with 5000 CAs and 10 ROAs per CA.
+            //
+            // So, we do this in memory. This should not be an issue for our application because
+            // we only read large files (snapshots) during startup. After this data is kept in
+            // memory. I.e. we do not expect that large files are read concurrently resulting
+            // in high memory consumption.
+            //
+            // Furthermore, this has a negligible impact on installations with a single small
+            // CA instance. And large operations - i.e. a bug Publication Server or Parent CA
+            // to lots of children, can afford a server with a decent amount of memory.
```

Comment:
With a minimal bump to MSRV from 1.56 to 1.57 you could use [Vec::try_reserve_exact](https://doc.rust-lang.org/std/vec/struct.Vec.html#method.try_reserve_exact) to attempt to pre-allocate the required amount of memory and fall back to the old method if insufficient memory is available. Is this worth considering or would you then consider that panic'ing is the right thing to do anyway as the server will need that memory to operate normally anyway?

Reply:
I think that the memory usage is minimal for small setups so it does not matter. And organisations who run a big parent CA, or more likely Publication Server, can allocate enough memory. I think panicking is probably right in this case.